### PR TITLE
[Beat] Add New Innovation Face Occlusion & Attribute

### PR DESCRIPTION
**Pull request by sonadztux**
## Type of changes

 

Please check the type of change your PR introduces:
- [ ] Bugfix
- [x] Enhancement
- [x] Code style update (formatting, renaming)
- [ ] Refactoring (no functional changes, no api changes)
- [ ] Build related changes
- [ ] Documentation content changes
- [ ] Other (please describe): 

## What is the current behaviour?

Belum ada backend handler untuk request face occlusion & attribute yang butuh 3 analitik sekaligus: face detection, face occlusion, dan face attribute.

## What is the new behavior?

Menambahkan backend handler untuk request face occlusion & attribute dengan mekanisme:
- Menerima satu input image dalam bentuk base64 string
- Lalu mengirim request ke face detection api yang mengembalikan 2 base64 string yang akan digunakan sebagai input ke face occlusion & attribute
- Kemudian mengirim request ke API face occlusion dan face attribute secara konkurensi
- Wrapping JSON response result dari API face occlusion dan face attribute menjadi satu JSON response result di backend weber
- Time consumption kurang lebih 1s
Screenshot at 2021-12-01 12-54-47

## Checklist

- [ ] Lint and unit tests pass locally with my changes
- [ ] I have added tests that prove my fix is effective or that my feature works
- [ ] I have added necessary documentation (if appropriate)
- [ ] Any dependent changes have been merged and published in downstream modules
- [x] I have filled the `[Unreleased]` [Changelog](https://github.com/nodefluxio/weber/blob/main/CHANGELOG.md

## Security Pentester API Checklist
- [ ] IDOR
- [ ] Malicious Script Injection
- [ ] SQL Injection

## Does this introduce a breaking change?

- [ ] Yes
- [x] No

## Other information

cc @gregoriusjimmy sample success JSON response result:
```json
{
    "message": "Service demo request success",
    "ok": true,
    "service_data": {
        "face_occlusion": {
            "job": {
                "result": {
                    "analytic_type": "FACE_OCCLUSION",
                    "result": [
                        {
                            "face_occlusion": {
                                "detections": [
                                    {
                                        "Forehead": {
                                            "bounding_box": {
                                                "height": 0.26785714285714285,
                                                "left": 0.20535714285714285,
                                                "top": 0.07142857142857142,
                                                "width": 0.5535714285714286
                                            },
                                            "confidence": 96.71
                                        }
                                    },
                                    {
                                        "Eye Right": {
                                            "bounding_box": {
                                                "height": 0.11607142857142858,
                                                "left": 0.22321428571428573,
                                                "top": 0.4017857142857143,
                                                "width": 0.21428571428571427
                                            },
                                            "confidence": 99.96
                                        }
                                    },
                                    {
                                        "Eye Left": {
                                            "bounding_box": {
                                                "height": 0.09821428571428571,
                                                "left": 0.5714285714285714,
                                                "top": 0.4107142857142857,
                                                "width": 0.19642857142857142
                                            },
                                            "confidence": 99.98
                                        }
                                    }
                                ],
                                "inference_time_seconds": 0.030247926712036133,
                                "occlusion": true
                            }
                        }
                    ],
                    "status": "success"
                }
            },
            "message": "Your cheek right, mouth, nose, eyebrow left, eyebrow right, cheek left, chin are occluded!",
            "ok": true
        },
        "face_attribute": {
            "job": {
                "result": {
                    "analytic_type": "FACE_ATTRIBUTE",
                    "result": [
                        {
                            "face_attribute": {
                                "additional_info_object": [],
                                "attribute_detected": true,
                                "occlusion_object": [
                                    {
                                        "confidence": "1.0",
                                        "label": "masker"
                                    }
                                ]
                            }
                        }
                    ],
                    "status": "success"
                }
            },
            "message": "Face Attribute Success",
            "ok": true
        }
    }
}
```

### CODE REVIEW

File: backend/controllers/request.go [hunk 1]
```diff
@@ -55,13 +55,19 @@ func RequestToServiceAnalytics(ctx *gin.Context, service models.Service, inputDa
...
+	var serviceData models.ServiceRequestResultData
 
 	requestData := inputData.Data
 	additionalParams, _ := json.Marshal(requestData.AdditionalParams)
 	postBody := []byte(fmt.Sprintf(`{ "additional_params": %v , "images":  [ "%v" ]}`,
 		string(additionalParams), strings.Join(requestData.Images, `", "`)))
 
-	serviceData, err := RequestToInnovationSync(postBody, service.Slug)
+	if service.Slug == "face-occlusion-attribute" {
+		ImplementFOAInnovation(ctx, postBody)
```
Comment (krismp) [inline backend/controllers/request.go:66]: `ImplementFOAInnovation` ini ngapain @sonadztux ?

Comment (sonadztux) [reply backend/controllers/request.go:66]: @krismp request ke 3 analitik: pertama ke analitik face detection dulu yang balikin dua image base64 string, terus request ke analitik face occlusion sama face attribute secara konkurensi.

Comment (krismp) [reply backend/controllers/request.go:66]: face occlusion sama face attribute bisa running independently?

Comment (krismp) [reply backend/controllers/request.go:66]: nama method nya apa ya, jgn `ImplementFOAInnovation` FOA itu apa ya?

Comment (sonadztux) [reply backend/controllers/request.go:66]: response result dari face detection gini:
```json
"face_detection": {
    "face_detected": true,
    "multiple_faces_detected": true,
    "result": {
        "input_face_attribute": [
            "data:image/jpeg;base64, /9j/4AAQSkZJRgABAQAA"
        ],
        "input_face_occlusion": [
          "data:image/jpeg;base64, /9j/4AAQSkZJRgABAQAAAQ"
        ]
    }
}
```

Comment (sonadztux) [reply backend/controllers/request.go:66]: > face occlusion sama face attribute bisa running independently?

bisa mas

> nama method nya apa ya, jgn ImplementFOAInnovation FOA itu apa ya?

FOA = Face Occlusion Attribute ehe

Comment (sonadztux) [reply backend/controllers/request.go:66]: `RequestToFaceOcclusionAttribute` wdyt?

Comment (krismp) [reply backend/controllers/request.go:66]: > RequestToFaceOcclusionAttribute wdyt?

Ok nih

Comment (sonadztux) [reply backend/controllers/request.go]: done https://github.com/nodefluxio/weber/pull/177/commits/ad923a9ebb6f73e8e0a16dcaf92365c81a7a11ae

File: backend/controllers/innovation.go [hunk 2]
```diff
@@ -5,13 +5,19 @@ import (
...
-
 	"net/http"
 	"os"
 
+	"github.com/gin-gonic/gin"
 	log "github.com/sirupsen/logrus"
+	"github.com/tidwall/gjson"
 )
 
+type foaServiceData struct {
```
Comment (krismp) [inline backend/controllers/innovation.go:16]: foa kaya gini di jabarin aja, gpp panjang

File: backend/controllers/innovation.go [hunk 3]
```diff
@@ -59,3 +65,122 @@ func RequestToInnovationSync(postBody []byte, innovationSlug string) (models.Ser
...
+
+	// Get image base64 string for input Face Occlusion & Attribute
+	inputFaceOcclusion := gjson.Get(string(resultFaceDetectionJson), "job.result.result.0.face_detection.result.input_face_occlusion.0").String()
+	inputFaceAttribute := gjson.Get(string(resultFaceDetectionJson), "job.result.result.0.face_detection.result.input_face_attribute.0").String()
+	
+	// Declare channels
+	resultFaceOcclusionChannel := make(chan models.ServiceRequestResultData)
+	resultFaceAttributeChannel := make(chan models.ServiceRequestResultData)
+	errorFaceOcclusionChannel := make(chan error)
+	errorFaceAttributeChannel := make(chan error)
```
Comment (krismp) [inline backend/controllers/innovation.go:123]: error nya gak dibikin satu channel aja?

File: backend/controllers/innovation.go [hunk 4]
```diff
@@ -59,3 +65,122 @@ func RequestToInnovationSync(postBody []byte, innovationSlug string) (models.Ser
...
+		case resultFaceAttribute := <-resultFaceAttributeChannel:
+			serviceData.FaceAttribute = resultFaceAttribute
+		case errorFaceOcclusion := <-errorFaceOcclusionChannel:
+			err = errorFaceOcclusion
+		case errorFaceAttribute := <-errorFaceAttributeChannel:
+			err = errorFaceAttribute
+		}
+	}
+
+	if err != nil {
```
Comment (krismp) [inline backend/controllers/innovation.go:155]: karena error disini cuma 1 yg di ambil, dan ini gak tw error milik siapa

Comment (sonadztux) [reply backend/controllers/innovation.go:155]: komen ini masih nyambung sama yang komen buat error channel jadiin satu aja ya, mas?

Comment (krismp) [reply backend/controllers/innovation.go:154]: yup

File: backend/controllers/innovation.go (hunk 2 above)
Comment (sonadztux) [reply backend/controllers/innovation.go]: done https://github.com/nodefluxio/weber/pull/177/commits/9e55e07bc8904df140e5ce8a473b4278c866786f

File: backend/controllers/innovation.go (hunk 3 above)
Comment (sonadztux) [reply backend/controllers/innovation.go]: done https://github.com/nodefluxio/weber/pull/177/commits/59b20b477dccf9693a0a1d6460c3845b486ca027

File: backend/controllers/innovation.go (hunk 4 above)
Comment (sonadztux) [reply backend/controllers/innovation.go:154]: ok done https://github.com/nodefluxio/weber/pull/177/commits/59b20b477dccf9693a0a1d6460c3845b486ca027

File: backend/controllers/innovation.go [hunk 5]
```diff
@@ -59,3 +65,108 @@ func RequestToInnovationSync(postBody []byte, innovationSlug string) (models.Ser
...
+			"message": err.Error(),
+		})
+		return
+	}
+
+	resultFaceDetectionJson, err := json.Marshal(resultFaceDetection)
+	if err != nil {
+		log.WithFields(log.Fields{
+			"error":     	err,
+			"json_result": 	resultFaceDetectionJson,
```
Comment (krismp) [inline backend/controllers/innovation.go:84]: ini `byte` kan? ke print di log nya gimana? perlu di `string(resultFaceDetectionJson)` gak? klo udh readable gpp

Comment (sonadztux) [reply backend/controllers/innovation.go:84]: @krismp oh iya itu `byte`, sorry kelewat. Nice catch, thank you, Mas.

Comment (sonadztux) [reply backend/controllers/innovation.go]: done https://github.com/nodefluxio/weber/pull/177/commits/703bc50d18b7798412092919cf5b9b2c51fbaa97

**Review by krismp — APPROVED** [review]

File: backend/database/seed.go [hunk 6]
```diff
@@ -200,6 +200,20 @@ func seedService(model *models.Model) {
 			CreatedAt:          time.Now(),
 			UpdatedAt:          time.Now(),
 		},
+		{
+			Name:               "Face Occlusion & Attribute",
```
Comment (lazuardyk) [inline backend/database/seed.go:204]: ini kalo jadi "Face Occlusion and Attribute Detection" gimana?

**Review by lazuardyk — APPROVED** [review]
Sisanya works. 2-3s, mantab

File: backend/database/seed.go (hunk 6 above)
Comment (sonadztux) [reply backend/database/seed.go:204]: di figma produk emang gitu.

Comment (lazuardyk) [reply backend/database/seed.go:204]: oalah yowes

**Pull request merged**